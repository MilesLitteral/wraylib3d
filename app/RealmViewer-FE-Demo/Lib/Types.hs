module Lib.Types where

type BooksWenv = WidgetEnv  BooksModel BooksEvt
type BooksNode = WidgetNode BooksModel BooksEvt
    
keyListType c v =
    case c of
      XBOX_CONTROLLER     -> xboxKeyList    v
      PS_CONTROLLER       -> psKeyList      v
      TP_CONTROLLER       -> thirdPartyList v 
      GENERIC_CONTROLLER  -> genericKeyList v

keyTextListType c =
  case c of
    XBOX_CONTROLLER     -> xboxKeyTextList
    PS_CONTROLLER       -> psKeyTextList    
    TP_CONTROLLER       -> thirdKeyTextList 
    GENERIC_CONTROLLER  -> thirdKeyTextList

xboxKeyTextList = [Mapping,        
  START,           
  SELECT,          
  L_DPAD,          
  R_DPAD,          
  U_DPAD,          
  D_DPAD,          
  L_STICK,         
  R_STICK,         
  BUTTON_A, 
  BUTTON_B,   
  BUTTON_X,     
  BUTTON_Y,  
  BUTTON_L1, 
  BUTTON_R1,
  BUTTON_L2, 
  BUTTON_R2,
  BUTTON_L3,
  BUTTON_R3
  ]       

psKeyTextList = [
    Mapping,           
    OPTION,          
    SHARE,           
    HOME_BUTTON,     
    TRACKPAD,         --Playstation Only
    PS_BUTTON_A,     
    PS_BUTTON_B,
    PS_BUTTON_X,     
    PS_BUTTON_Y,     
    PS_BUTTON_L1,    
    PS_BUTTON_R1,    
    PS_BUTTON_L2,    
    PS_BUTTON_R2,    
    PS_BUTTON_L3,    
    PS_BUTTON_R3   
    ] 

thirdKeyTextList =  [
    Mapping,                 
    KEYBOARD_KEY,    
    L_MOUSE_KEY,     
    R_MOUSE_KEY,     
    C_MOUSE_KEY,     
    XR_GESTURE    
    ]    

xboxKeyList c = 
  case c of
    Mapping         -> image "./assets/images/ControllerIcons/mapping_icon.png"                   `styleBasic` [height 15,width 15]
    START           -> image "./assets/images/ControllerIcons/Special/Menu_dark.png"              `styleBasic` [height 15,width 15]
    SELECT          -> image "./assets/images/ControllerIcons/Special/View_dark.png"              `styleBasic` [height 15,width 15]
    L_DPAD          -> image "./assets/images/ControllerIcons/Digipad/Digipad_left_dark.png"      `styleBasic` [height 15,width 15]
    R_DPAD          -> image "./assets/images/ControllerIcons/Digipad/Digipad_right_dark.png"     `styleBasic` [height 15,width 15]
    U_DPAD          -> image "./assets/images/ControllerIcons/Digipad/Digipad_up_dark.png"        `styleBasic` [height 15,width 15]
    D_DPAD          -> image "./assets/images/ControllerIcons/Digipad/Digipad_down_dark.png"      `styleBasic` [height 15,width 15]
    L_STICK         -> image "./assets/images/ControllerIcons/Stick/L_dark.png"      `styleBasic` [height 15,width 15]--, sizeReqW $ fixedSize 25, sizeReqH $ fixedSize 25]
    R_STICK         -> image "./assets/images/ControllerIcons/Stick/R_dark.png"      `styleBasic` [height 15,width 15]--, sizeReqW $ fixedSize 25, sizeReqH $ fixedSize 25]
    BUTTON_A        -> image "./assets/images/ControllerIcons/Button/A_color_dark.png"      `styleBasic` [height 15,width 15]--, sizeReqW $ fixedSize 25, sizeReqH $ fixedSize 25]
    BUTTON_B        -> image "./assets/images/ControllerIcons/Button/B_color_dark.png"      `styleBasic` [height 15,width 15]--, sizeReqW $ fixedSize 25, sizeReqH $ fixedSize 25]
    BUTTON_X        -> image "./assets/images/ControllerIcons/Button/X_color_dark.png"      `styleBasic` [height 15,width 15]--, sizeReqW $ fixedSize 25, sizeReqH $ fixedSize 25]
    BUTTON_Y        -> image "./assets/images/ControllerIcons/Button/Y_color_dark.png"      `styleBasic` [height 15,width 15]
    BUTTON_L1       -> image "./assets/images/ControllerIcons/ShoulderButton/LB_dark.png"      `styleBasic` [height 15,width 15]
    BUTTON_R1       -> image "./assets/images/ControllerIcons/ShoulderButton/RB_dark.png"      `styleBasic` [height 15,width 15]
    BUTTON_L2       -> image "./assets/images/ControllerIcons/ShoulderButton/LT_dark.png"      `styleBasic` [height 15,width 15]
    BUTTON_R2       -> image "./assets/images/ControllerIcons/ShoulderButton/RT_dark.png"      `styleBasic` [height 15,width 15]
    BUTTON_L3       -> image "./assets/images/ControllerIcons/ShoulderButton/windows_sm.png"      `styleBasic` [height 15,width 15]
    BUTTON_R3       -> image "./assets/images/ControllerIcons/ShoulderButton/windows_sm.png"      `styleBasic` [height 15,width 15]

psKeyList c = 
  case c of
    Mapping         -> image "./assets/images/ControllerIcons/mapping_icon.png"                   `styleBasic` [sizeReqW $ fixedSize 50, sizeReqH $ fixedSize 50]
    OPTION          -> image "./assets/images/ControllerIcons/PlaystationButtons/Share.png"       `styleBasic` [sizeReqW $ fixedSize 50, sizeReqH $ fixedSize 50]
    SHARE           -> image "./assets/images/ControllerIcons/PlaystationButtons/Share.png"       `styleBasic` [sizeReqW $ fixedSize 50, sizeReqH $ fixedSize 50]
    HOME_BUTTON     -> image "./assets/images/ControllerIcons/PlaystationButtons/PSButton.png"    `styleBasic` [sizeReqW $ fixedSize 50, sizeReqH $ fixedSize 50]
    TRACKPAD        -> image "./assets/images/ControllerIcons/PlaystationButtons/Touchbar.png"    `styleBasic` [sizeReqW $ fixedSize 50, sizeReqH $ fixedSize 50] --Playstation Only
    PS_BUTTON_A     -> image "./assets/images/ControllerIcons/PlaystationButtons/X.png"      `styleBasic` [sizeReqW $ fixedSize 50, sizeReqH $ fixedSize 50]
    PS_BUTTON_B     -> image "./assets/images/ControllerIcons/PlaystationButtons/Circle.png"      `styleBasic` [sizeReqW $ fixedSize 50, sizeReqH $ fixedSize 50]
    PS_BUTTON_X     -> image "./assets/images/ControllerIcons/PlaystationButtons/Square.png"      `styleBasic` [sizeReqW $ fixedSize 50, sizeReqH $ fixedSize 50]
    PS_BUTTON_Y     -> image "./assets/images/ControllerIcons/PlaystationButtons/Triangle.png"      `styleBasic` [sizeReqW $ fixedSize 50, sizeReqH $ fixedSize 50]
    PS_BUTTON_L1    -> image "./assets/images/ControllerIcons/PlaystationButtons/L1.png"      `styleBasic` [sizeReqW $ fixedSize 50, sizeReqH $ fixedSize 50]
    PS_BUTTON_R1    -> image "./assets/images/ControllerIcons/PlaystationButtons/R1.png"      `styleBasic` [sizeReqW $ fixedSize 50, sizeReqH $ fixedSize 50]
    PS_BUTTON_L2    -> image "./assets/images/ControllerIcons/PlaystationButtons/L2.png"      `styleBasic` [sizeReqW $ fixedSize 50, sizeReqH $ fixedSize 50]
    PS_BUTTON_R2    -> image "./assets/images/ControllerIcons/PlaystationButtons/R2.png"      `styleBasic` [sizeReqW $ fixedSize 50, sizeReqH $ fixedSize 50]
    PS_BUTTON_L3    -> image "./assets/images/ControllerIcons/PlaystationButtons/L3.png"      `styleBasic` [sizeReqW $ fixedSize 50, sizeReqH $ fixedSize 50]
    PS_BUTTON_R3    -> image "./assets/images/ControllerIcons/PlaystationButtons/R3.png"      `styleBasic` [sizeReqW $ fixedSize 50, sizeReqH $ fixedSize 50]

thirdPartyList c =   
  case c of
    Mapping         -> image "./assets/images/ControllerIcons/mapping_icon.png"       `styleBasic` [sizeReqW $ fixedSize 50, sizeReqH $ fixedSize 50]        
    KEYBOARD_KEY    -> image "./assets/images/ControllerIcons/keyboard_key.png"       `styleBasic` [sizeReqW $ fixedSize 50, sizeReqH $ fixedSize 50]
    L_MOUSE_KEY     -> image "./assets/images/ControllerIcons/mouse_left_click.png"   `styleBasic` [sizeReqW $ fixedSize 50, sizeReqH $ fixedSize 50]
    R_MOUSE_KEY     -> image "./assets/images/ControllerIcons/mouse_right_click.png"  `styleBasic` [sizeReqW $ fixedSize 50, sizeReqH $ fixedSize 50]
    C_MOUSE_KEY     -> image "./assets/images/ControllerIcons/mouse_middle_click.png" `styleBasic` [sizeReqW $ fixedSize 50, sizeReqH $ fixedSize 50]
    XR_GESTURE      -> image "./assets/images/ControllerIcons/xr_gesture.png"         `styleBasic` [sizeReqW $ fixedSize 50, sizeReqH $ fixedSize 50]  

genericKeyList c =   
  case c of
    Mapping         -> image "./assets/images/ControllerIcons/mapping_icon.png"       `styleBasic` [sizeReqW $ fixedSize 50, sizeReqH $ fixedSize 50]        
    KEYBOARD_KEY    -> image "./assets/images/ControllerIcons/keyboard_key.png"       `styleBasic` [sizeReqW $ fixedSize 50, sizeReqH $ fixedSize 50]
    L_MOUSE_KEY     -> image "./assets/images/ControllerIcons/mouse_left_click.png"   `styleBasic` [sizeReqW $ fixedSize 50, sizeReqH $ fixedSize 50]
    R_MOUSE_KEY     -> image "./assets/images/ControllerIcons/mouse_right_click.png"  `styleBasic` [sizeReqW $ fixedSize 50, sizeReqH $ fixedSize 50]
    C_MOUSE_KEY     -> image "./assets/images/ControllerIcons/mouse_middle_click.png" `styleBasic` [sizeReqW $ fixedSize 50, sizeReqH $ fixedSize 50]
    XR_GESTURE      -> image "./assets/images/ControllerIcons/xr_gesture.png"         `styleBasic` [sizeReqW $ fixedSize 50, sizeReqH $ fixedSize 50]  

appModel :: BooksModel
appModel   = initialModel

doneBg :: Color
doneBg     = rgbHex "#CFF6E2"

doneFg :: Color
doneFg     = rgbHex "#459562"

pendingBg :: Color
pendingBg  = rgbHex "#F5F0CC"

pendingFg :: Color
pendingFg  = rgbHex "#827330"

grayLight :: Color
grayLight  = rgbHex "#9E9E9E"

grayDark  :: Color
grayDark   = rgbHex "#393939"

grayDarker :: Color
grayDarker = rgbHex "#2E2E2E"

customLightTheme :: Theme
customLightTheme = lightTheme
  & L.userColorMap . at "rowBgColor" ?~ rgbHex "#ECECEC"

customDarkTheme :: Theme
customDarkTheme = darkTheme
  & L.userColorMap . at "rowBgColor" ?~ rgbHex "#656565"
  