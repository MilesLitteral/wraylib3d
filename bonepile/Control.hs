{-# LANGUAGE AllowAmbiguousTypes #-}

module Manifest.Control where

import Manifest.Control.NetworkManager
import Manifest.Control.UDP
