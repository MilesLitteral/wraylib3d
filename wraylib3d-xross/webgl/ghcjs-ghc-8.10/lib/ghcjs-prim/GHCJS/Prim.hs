{-# LANGUAGE MagicHash, DeriveDataTypeable, ScopedTypeVariables, CPP #-}
#ifdef __GHCJS__
{-# LANGUAGE JavaScriptFFI,
             GHCForeignImportPrim,
             UnliftedFFITypes,
             UnboxedTuples
  #-}
#endif

module GHCJS.Prim ( JSVal(..), JSVal#
                  , JSException(..)
                  , WouldBlockException(..)
#ifdef ghcjs_HOST_OS
                  , toIO
                  , resolve
                  , resolveIO
                  , mkJSException
                  , fromJSString
                  , toJSString
                  , toJSArray
                  , fromJSArray
                  , fromJSInt
                  , toJSInt
                  , isNull
                  , isUndefined
                  , jsNull
                  , getProp
                  , getProp'
                  , getProp#
                  , unsafeGetProp
                  , unsafeGetProp'
                  , unsafeGetProp#
                  , unpackJSString#
                  , unpackJSStringUtf8#
                  , unsafeUnpackJSString#
                  , unsafeUnpackJSStringUtf8#
                  , unpackJSStringUtf8##
                  , unsafeUnpackJSStringUtf8##

#endif
                  ) where

import           Data.Typeable (Typeable)
import           Unsafe.Coerce (unsafeCoerce)

import           GHC.Prim
import qualified GHC.Exception as Ex
import qualified GHC.Exts as Exts
import qualified GHC.CString as GHC
import           GHC.IO

{-
  JSVal is a boxed type that can be used as FFI
  argument or result.
-}
#ifdef ghcjs_HOST_OS
data JSVal  = JSVal ByteArray#
type JSVal# = ByteArray#
#else
data JSVal  = JSVal Addr#
type JSVal# = Addr#
#endif

{-
  When a JavaScript exception is raised inside
  a safe or interruptible foreign call, it is converted
  to a JSException
 -}
data JSException = JSException JSVal String
  deriving (Typeable)

instance Ex.Exception JSException

instance Show JSException where
  show (JSException _ xs) = "JavaScript exception: " ++ xs

#ifdef ghcjs_HOST_OS

{-# NOINLINE toIO #-}
toIO :: Exts.Any -> IO Exts.Any
toIO x = pure x

{-# NOINLINE resolve #-}
resolve :: JSVal# -> JSVal# -> Exts.Any -> IO ()
resolve accept reject x = resolveIO accept reject (pure x)

{-# NOINLINE resolveIO #-}
resolveIO :: JSVal# -> JSVal# -> IO Exts.Any -> IO ()
resolveIO accept reject x =
  (x >>= evaluate >>= js_callback_any accept) `catch`
  (\(e::Ex.SomeException) -> do
    exceptionText <- evaluate (toJSString $ Ex.displayException e) `catch`
                         (\(_::Ex.SomeException) -> evaluate (toJSString "unknown exception"))
    js_callback_jsval reject exceptionText)

mkJSException :: JSVal -> IO JSException
mkJSException ref =
  return (JSException (unsafeCoerce ref) (fromJSString ref))

{- | Low-level conversion utilities for packages that cannot
     depend on ghcjs-base
 -}

{- | returns an empty string if the JSVal does not contain
     a string
 -}
fromJSString :: JSVal -> String
fromJSString = unsafeCoerce . js_fromJSString
{-# INLINE fromJSString #-}

toJSString :: String -> JSVal
toJSString = js_toJSString . unsafeCoerce . seqList
{-# INLINE [0] toJSString #-}
{-# RULES
"GHCJSPRIM toJSString/literal" forall a.
  toJSString (GHC.unpackCString# a) = JSVal (unsafeUnpackJSStringUtf8## a)
"GHCJSPRIM toJSString/literalUtf8" forall a.
  toJSString (GHC.unpackCStringUtf8# a) = JSVal (unsafeUnpackJSStringUtf8## a)
  #-}

fromJSArray :: JSVal -> IO [JSVal]
fromJSArray = unsafeCoerce . js_fromJSArray
{-# INLINE fromJSArray #-}

toJSArray :: [JSVal] -> IO JSVal
toJSArray = js_toJSArray . unsafeCoerce . seqList
{-# INLINE toJSArray #-}

{- | returns zero if the JSVal does not contain a number
 -}
fromJSInt :: JSVal -> Int
fromJSInt = js_fromJSInt
{-# INLINE fromJSInt #-}

toJSInt :: Int -> JSVal
toJSInt = js_toJSInt
{-# INLINE toJSInt #-}

isNull :: JSVal -> Bool
isNull = js_isNull
{-# INLINE isNull #-}

isUndefined :: JSVal -> Bool
isUndefined = js_isUndefined
{-# INLINE isUndefined #-}

jsNull :: JSVal
jsNull = js_null
{-# INLINE CONLIKE jsNull #-}

getProp :: JSVal -> String -> IO JSVal
getProp o p = js_getProp o (unsafeCoerce $ seqList p)
{-# INLINE [0] getProp #-}
{-# RULES
"GHCJSPRIM getProp/literal" forall o a.
  getProp o (GHC.unpackCString# a) = getProp# o a
"GHCJSPRIM getProp/literalUtf8" forall o a.
  getProp o (GHC.unpackCStringUtf8# a) = getPropUtf8# o a
  #-}

-- | only safe on immutable object
unsafeGetProp :: JSVal -> String -> JSVal
unsafeGetProp o p = js_unsafeGetProp o (unsafeCoerce $ seqList p)
{-# INLINE [0] unsafeGetProp #-}
{-# RULES
"GHCJSPRIM unsafeGetProp/literal" forall o a.
  unsafeGetProp o (GHC.unpackCString# a) = unsafeGetProp# o a
"GHCJSPRIM unsafeGetProp/literalUtf8" forall o a.
  unsafeGetProp o (GHC.unpackCStringUtf8# a) = unsafeGetPropUtf8# o a
  #-}

getProp' :: JSVal -> JSVal -> IO JSVal
getProp' o p = js_getProp' o p
{-# INLINE [0] getProp' #-}
{-# RULES
"GHCJSPRIM getProp'/literal" forall o a.
  getProp' o (unsafeUnpackJSString# a) = getProp# o a
"GHCJSPRIM getProp'/literalUtf8" forall o a.
  getProp' o (unsafeUnpackJSStringUtf8# a) = getPropUtf8# o a
  #-}

-- | only safe on immutable object
unsafeGetProp' :: JSVal -> JSVal -> JSVal
unsafeGetProp' o p = js_unsafeGetProp' o p
{-# INLINE [0] unsafeGetProp' #-}
{-# RULES
"GHCJSPRIM unsafeGetProp'/literal" forall o a.
  unsafeGetProp' o (unsafeUnpackJSString# a) = unsafeGetPropUtf8# o a
"GHCJSPRIM unsafeGetProp'/literalUtf8" forall o a.
  unsafeGetProp' o (unsafeUnpackJSStringUtf8# a) = unsafeGetPropUtf8# o a
  #-}


-- | only safe on immutable Addr#
getProp# :: JSVal -> Addr# -> IO JSVal
getProp# (JSVal o) p = IO $
  \s -> case getPropUtf8## o p s of (# s', v #) -> (# s', JSVal v #)
{-# INLINE [0] getProp# #-}
-- js_getProp# o p

-- | only safe on immutable Addr#
getPropUtf8# :: JSVal -> Addr# -> IO JSVal
getPropUtf8# (JSVal o) p = IO $
  \s -> case getPropUtf8## o p s of (# s', v #) -> (# s', JSVal v #)
{-# INLINE [0] getPropUtf8# #-}

getPropUtf8## :: JSVal# -> Addr# -> State# s -> (# State# s, JSVal# #)
getPropUtf8## o p = js_getPropUtf8## o p
{-# NOINLINE getPropUtf8## #-}

-- | only safe on immutable Addr# and JSVal
unsafeGetProp# :: JSVal -> Addr# -> JSVal
unsafeGetProp# (JSVal o) p = JSVal (unsafeGetPropUtf8## o p)
{-# INLINE [0] unsafeGetProp# #-}

-- | only safe on immutable Addr# and JSVal
unsafeGetPropUtf8# :: JSVal -> Addr# -> JSVal
unsafeGetPropUtf8# (JSVal o) p = JSVal (unsafeGetPropUtf8## o p)
{-# INLINE [0] unsafeGetPropUtf8# #-}

unsafeGetPropUtf8## :: JSVal# -> Addr# -> JSVal#
unsafeGetPropUtf8## o p = js_unsafeGetPropUtf8## o p
{-# NOINLINE unsafeGetPropUtf8## #-}

unpackJSString# :: Addr# -> IO JSVal
unpackJSString# a = IO $
  \s -> case unpackJSStringUtf8## a s of (# s', v #) -> (# s', JSVal v #)
{-# INLINE [0] unpackJSString# #-}

unpackJSStringUtf8# :: Addr# -> IO JSVal
unpackJSStringUtf8# a = IO $
  \s -> case unpackJSStringUtf8## a s of (# s', v #) -> (# s', JSVal v #)
{-# INLINE [0] unpackJSStringUtf8# #-}

unpackJSStringUtf8## :: Addr# -> State# s -> (# State# s, JSVal# #)
unpackJSStringUtf8## a s = js_unpackJSStringUtf8## a s
{-# NOINLINE unpackJSStringUtf8## #-}

-- | only safe on immutable Addr#
unsafeUnpackJSString# :: Addr# -> JSVal
unsafeUnpackJSString# a = JSVal (unsafeUnpackJSStringUtf8## a)
  -- js_unsafeUnpackJSString# a
{-# INLINE [0] unsafeUnpackJSString# #-}

-- | only safe on immutable Addr#
unsafeUnpackJSStringUtf8# :: Addr# -> JSVal
unsafeUnpackJSStringUtf8# a = JSVal (unsafeUnpackJSStringUtf8## a)
{-# INLINE [0] unsafeUnpackJSStringUtf8# #-}

unsafeUnpackJSStringUtf8## :: Addr# -> JSVal#
unsafeUnpackJSStringUtf8## a = js_unsafeUnpackJSStringUtf8## a
{-# NOINLINE unsafeUnpackJSStringUtf8## #-}


-- reduce the spine and all list elements to whnf
seqList :: [a] -> [a]
seqList xs = go xs `seq` xs
  where go (x:xs) = x `seq` go xs
        go []     = ()

seqListSpine :: [a] -> [a]
seqListSpine xs = go xs `seq` xs
  where go (x:xs) = go xs
        go []     = ()

foreign import javascript unsafe "h$toHsString($1)"
  js_fromJSString :: JSVal -> Exts.Any

foreign import javascript unsafe "h$fromHsString($1)"
  js_toJSString :: Exts.Any -> JSVal

foreign import javascript unsafe "h$toHsListJSVal($1)"
  js_fromJSArray :: JSVal -> IO Exts.Any

foreign import javascript unsafe "h$fromHsListJSVal($1)"
  js_toJSArray :: Exts.Any -> IO JSVal

foreign import javascript unsafe "$1 === null"
  js_isNull :: JSVal -> Bool

foreign import javascript unsafe "$1 === undefined"
  js_isUndefined :: JSVal -> Bool

foreign import javascript unsafe "$r = typeof($1) === 'number' ? ($1|0) : 0;"
  js_fromJSInt :: JSVal -> Int

foreign import javascript unsafe "$r = $1;"
  js_toJSInt :: Int -> JSVal

foreign import javascript unsafe "$r = null;"
  js_null :: JSVal

foreign import javascript unsafe "$1[h$fromHsString($2)]"
  js_getProp :: JSVal -> Exts.Any -> IO JSVal

foreign import javascript unsafe "$1[h$fromHsString($2)]"
  js_unsafeGetProp :: JSVal -> Exts.Any -> JSVal

foreign import javascript unsafe "$1[$2]"
  js_getProp' :: JSVal -> JSVal -> IO JSVal

foreign import javascript unsafe "$1[$2]"
  js_unsafeGetProp' :: JSVal -> JSVal -> JSVal

foreign import javascript unsafe "$1[h$decodeUtf8z($2_1, $2_2)]"
  js_getPropUtf8## :: JSVal# -> Addr# -> State# s -> (# State# s, JSVal# #)

foreign import javascript unsafe "$1[h$decodeUtf8z($2_1, $2_2)]"
  js_unsafeGetPropUtf8## :: JSVal# -> Addr# -> JSVal#

foreign import javascript unsafe "h$decodeUtf8z($1_1, $1_2)"
  js_unpackJSStringUtf8## :: Addr# -> State# s -> (# State# s, JSVal# #)


foreign import javascript unsafe "h$decodeUtf8z($1_1, $1_2)"
  js_unsafeUnpackJSStringUtf8## :: Addr# -> JSVal#

foreign import javascript unsafe "$1($2)"
  js_callback_any :: JSVal# -> Exts.Any -> IO ()

foreign import javascript unsafe "$1($2)"
  js_callback_jsval :: JSVal# -> JSVal -> IO ()

#endif

{- | If a synchronous thread tries to do something that can only
     be done asynchronously, and the thread is set up to not
     continue asynchronously, it receives this exception.
 -}
data WouldBlockException = WouldBlockException
  deriving (Typeable)

instance Show WouldBlockException where
  show _ = "thread would block"

instance Ex.Exception WouldBlockException
