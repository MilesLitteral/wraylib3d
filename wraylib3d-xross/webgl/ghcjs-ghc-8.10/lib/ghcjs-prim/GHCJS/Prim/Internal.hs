{- | Code used by the RTS

 -}

module GHCJS.Prim.Internal ( blockedIndefinitelyOnMVar
                           , blockedIndefinitelyOnSTM
                           , wouldBlock
                           , ignoreException
                           , setCurrentThreadResultException
                           , setCurrentThreadResultValue
                           ) where

import           Control.Exception

import           GHCJS.Prim
import           GHC.Exts

wouldBlock :: SomeException
wouldBlock = toException WouldBlockException

blockedIndefinitelyOnMVar :: SomeException
blockedIndefinitelyOnMVar = toException BlockedIndefinitelyOnMVar

blockedIndefinitelyOnSTM :: SomeException
blockedIndefinitelyOnSTM = toException BlockedIndefinitelyOnSTM

ignoreException :: SomeException -> IO ()
ignoreException _ = return ()

setCurrentThreadResultException :: SomeException -> IO ()
setCurrentThreadResultException e
  | Just WouldBlockException <- fromException e =
      js_setCurrentThreadResultWouldBlock
  | Just (JSException v _) <- fromException e =
      js_setCurrentThreadResultJSException v
  | otherwise =
      js_setCurrentThreadResultHaskellException (toJSString (show e))

setCurrentThreadResultValue :: IO JSVal -> IO ()
setCurrentThreadResultValue x = js_setCurrentThreadResultValue =<< x

foreign import javascript unsafe
  "h$setCurrentThreadResultWouldBlock();"
  js_setCurrentThreadResultWouldBlock :: IO ()

foreign import javascript unsafe
  "h$setCurrentThreadResultJSException($1);"
  js_setCurrentThreadResultJSException :: JSVal -> IO ()

foreign import javascript unsafe
  "h$setCurrentThreadResultHaskellException($1);"
  js_setCurrentThreadResultHaskellException :: JSVal -> IO ()

foreign import javascript unsafe
  "h$setCurrentThreadResultValue($1);"
  js_setCurrentThreadResultValue :: JSVal -> IO ()

