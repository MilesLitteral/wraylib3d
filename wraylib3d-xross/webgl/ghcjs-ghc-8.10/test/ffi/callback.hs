{-# LANGUAGE JavaScriptFFI, OverloadedStrings, ScopedTypeVariables #-}

module Main where

import Data.Char
import Data.IORef
import Data.Monoid

import Control.Concurrent
import Control.Concurrent.MVar
import Control.Exception
import Control.Monad

import GHCJS.Foreign
import GHCJS.Foreign.Callback
import GHCJS.Types
import GHCJS.Prim

import GHCJS.Marshal
import qualified Data.JSString as JSS
import qualified Data.JSString.Internal as JSSI

import System.Mem
import System.Mem.Weak

main :: IO ()
main = sequence_ [test1, test2, test3]

tou :: JSVal -> String
tou = map toUpper . JSS.unpack . jss
{-# NOINLINE tou #-}

-- sync callbacks without result
test1 :: IO ()
test1 = do
  js_log "test1"
  mv <- newEmptyMVar
  let jsa = jsval ("ab" :: JSString)
      jsb = jsval ("bc" :: JSString)
      jsc = jsval ("cd" :: JSString)
      x  = takeMVar mv >> js_log "mvar taken"
      x' = x `catch` \(_::WouldBlockException) -> js_log "would block"
      t0 n ob xx = t n (syncCallback ob xx) (js_runCallback n)
      t1 n ob xx = t n (syncCallback1 ob xx) (js_runCallback1 n jsa)
      t2 n ob xx = t n (syncCallback2 ob xx) (js_runCallback2 n jsa jsb)
      t3 n ob xx = t n (syncCallback3 ob xx) (js_runCallback3 n jsa jsb jsc)
      t name mkCB runCB = do
        js_log (">>> " <> name)
        cb <- mkCB
        runCB cb
        putMVar mv ()
        threadDelay 200000
        tryTakeMVar mv
        releaseCallback cb
  t0 "cb1" ContinueAsync   x  -- mvar taken async / result false
  t0 "cb2" ThrowWouldBlock x  -- result false
  t0 "cb3" ContinueAsync   x' -- mvar taken async / result false
  t0 "cb4" ThrowWouldBlock x' -- mvar not taken / would block printed / result true
  t1 "cb5" ThrowWouldBlock (js_log . JSS.pack . tou)
  t2 "cb6" ThrowWouldBlock (\a b -> js_log (JSS.pack $ tou a ++ tou b))
  t3 "cb7" ThrowWouldBlock (\a b c -> js_log (JSS.pack $ tou a ++ tou b ++ tou c))

-- sync callbacks with result
test2 :: IO ()
test2 = do
  js_log "test2"
  mv <- newEmptyMVar
  let jsa = jsval ("ab" :: JSString)
      jsb = jsval ("bc" :: JSString)
      jsc = jsval ("cd" :: JSString)
      s1       = return . jsval . JSS.pack . tou
      s2 a b   = return . jsval . JSS.pack $ tou a ++ tou b
      s3 a b c = return . jsval . JSS.pack $ tou a ++ tou b ++ tou c
      t0 n xx = t n (syncCallback' xx)  (js_runCallback n)
      t1 n xx = t n (syncCallback1' xx) (js_runCallback1 n jsa)
      t2 n xx = t n (syncCallback2' xx) (js_runCallback2 n jsa jsb)
      t3 n xx = t n (syncCallback3' xx) (js_runCallback3 n jsa jsb jsc)
      t name mkCB runCB = do
        js_log (">>> " <> name)
        cb <- mkCB
        runCB cb
        putMVar mv ()
        threadDelay 200000
        tryTakeMVar mv
        releaseCallback cb
      tmc = takeMVar mv `catch` \(_::WouldBlockException) -> js_log "would block"
  t1 "cb1" (return . jsval . JSS.pack . tou)
  t2 "cb2" (\a b -> return . jsval . JSS.pack $ tou a ++ tou b)
  t3 "cb3" (\a b c -> return . jsval . JSS.pack $ tou a ++ tou b ++ tou c)
  t1 "cb4" (\a     -> takeMVar mv >> s1 a)
  t2 "cb5" (\a b   -> takeMVar mv >> s2 a b)
  t3 "cb6" (\a b c -> takeMVar mv >> s3 a b c)
  t1 "cb7" (\a     -> tmc >> s1 a)
  t2 "cb8" (\a b   -> tmc >> s2 a b)
  t3 "cb9" (\a b c -> tmc >> s3 a b c)
  
test3 :: IO ()
test3 = do
  js_log "test3"
  ior <- newIORef "abc"
  w   <- mkWeakIORef ior (js_log "finalized")
  let v1 = jsval ("xyz" :: JSString)
  cb1 <- syncCallback  ThrowWouldBlock (readIORef ior >>= js_log . ("value: " <>))
  cb2 <- syncCallback1 ThrowWouldBlock
    (\x -> readIORef ior >>= \y -> js_log (jss x <> y))
  performGC
  js_runCallback  "cb1" cb1
  js_runCallback1 "cb2" v1 cb2 
  writeIORef ior "def"
  js_runCallback  "cb1" cb1
  js_runCallback1 "cb2" v1 cb2 
  js_log "test3.1"
  performGC
  threadDelay 500000
  js_log =<< maybe (return "<empty>") readIORef =<< deRefWeak w
  js_log "test3.2" -- finalizer should not have run
  releaseCallback cb1
  releaseCallback cb2
  performGC
  threadDelay 500000
  js_log =<< maybe (return "<empty>") readIORef =<< deRefWeak w
  js_log "test3.3" -- finalizer should have run

foreign import javascript unsafe
  "h$log($1);"
  js_log :: JSString -> IO ()

foreign import javascript unsafe
  "try { h$log($1 + ' result: ' + $2()) } catch(e) { h$log($1 + ' exception: ' + e); }"
  js_runCallback :: JSString -> Callback a -> IO ()

foreign import javascript unsafe
  "try { h$log($1 + ' result: ' + $3($2)) } catch(e) { h$log($1 + ' exception: ' + e); }"
  js_runCallback1 :: JSString -> JSVal -> Callback a -> IO ()

foreign import javascript unsafe
  "try { h$log($1 + ' result: ' + $4($2,$3)) } catch(e) { h$log($1 + ' exception: ' + e); }"
  js_runCallback2 :: JSString -> JSVal -> JSVal -> Callback a -> IO ()

foreign import javascript unsafe
  "try { h$log($1 + ' result: ' + $5($2,$3,$4)) } catch(e) { h$log($1 + ' exception: ' + e); }"
  js_runCallback3 :: JSString -> JSVal -> JSVal -> JSVal -> Callback a -> IO ()


foreign import javascript unsafe "$r = $1;" jss :: JSVal -> JSString
