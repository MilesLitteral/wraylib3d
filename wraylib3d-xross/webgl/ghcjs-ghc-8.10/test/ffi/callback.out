test1
>>> cb1
cb1 result: false
mvar taken
>>> cb2
cb2 result: false
>>> cb3
cb3 result: false
mvar taken
>>> cb4
would block
cb4 result: true
>>> cb5
AB
cb5 result: true
>>> cb6
ABBC
cb6 result: true
>>> cb7
ABBCCD
cb7 result: true
test2
>>> cb1
cb1 result: AB
>>> cb2
cb2 result: ABBC
>>> cb3
cb3 result: ABBCCD
>>> cb4
cb4 exception: Haskell Operation would block
>>> cb5
cb5 exception: Haskell Operation would block
>>> cb6
cb6 exception: Haskell Operation would block
>>> cb7
would block
cb7 result: AB
>>> cb8
would block
cb8 result: ABBC
>>> cb9
would block
cb9 result: ABBCCD
test3
value: abc
cb1 result: true
xyzabc
cb2 result: true
value: def
cb1 result: true
xyzdef
cb2 result: true
test3.1
def
test3.2
finalized
<empty>
test3.3
